<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/html4/loose.dtd">

<html>
  <head><title>JBrowse getting started guide</title></head>
  <link rel="stylesheet" type="text/css" href="../docs.css"></link>
  <body>
    <div class="container">
      <h2>Getting Started With JBrowse</h2>
      <p>This page should be enough to get you started.  If you run into trouble, please <a href="https://lists.sourceforge.net/lists/listinfo/gmod-ajax">join the mailing list</a> and ask us about it.</p>
      <p>
        <h4>Contents:</h4>
        <ul>
          <li><a href="#Prerequisites">Prerequisites</a></li>
          <li><a href="#Download_install">Download/install</a></li>
          <li><a href="#Reference_Sequences">Reference Sequences</a></li>
          <li><a href="#Features">Features</a></li>
          <li><a href="#Names">Names</a></li>
          <li><a href="#Quantitative_tracks">Quantitative tracks</a></li>
        </ul>
      </p>

      <a name="Prerequisites" id="Prerequisites"></a>
      <h3>Prerequisites</h3>
      <span class="sidebar">
        Alternatively, you can skip installing JBrowse and its prerequisites by launching a machine image using <a href="http://aws.amazon.com/ec2/">Amazon.com's EC2 service</a>.  We've prepared an image for use with EC2.  For more information, see <a href="http://biowiki.org/view/JBrowse/AmazonMachineImage">our wiki page</a>.
      </span>
      <p>
        From CPAN:
        <ul>
          <li><span class="code">BioPerl</span> <i>(version 1.6)</i></li>
          <li><span class="code">JSON</span> <i>(version 2)</i></li>
          <li><span class="code">JSON::XS</span> <i>(optional, for speed)</i></li>
        </ul>
      </p>

      <a name="Download_install" id="Download_install"></a>
      <h3>Download/install</h3>
      <p>
        <a href="http://github.com/jbrowse/jbrowse/tree/master">Get JBrowse from GitHub</a> (click on the "download" button)
      </p>
      <p>
        Then unzip/untar JBrowse in a directory that's served by apache (e.g., <span class="code">/var/www/html/</span>)
      </p>

      <a name="Reference_Sequences" id="Reference_Sequences"></a>
      <h3>Reference Sequences</h3>
      <p>
        Tell JBrowse about your reference sequences with <span class="code">bin/prepare-refseqs.pl</span>
      </p>
      <p>
        (JBrowse generates files with links that are relative to the directory where index.html is located, so you should run the JBrowse scripts from that directory)
        <pre>$ bin/prepare-refseqs.pl --fasta docs/tutorial/data_files/volvox.fa</pre>
      </p>

      <a name="Features" id="Features"></a>
      <h3>Features</h3>
      <p>
        There are two options for telling JBrowse about your feature data: if you have a genomic annotation database such as <a href="http://gmod.org/wiki/Chado">chado</a>, <a href="http://www.bioperl.org/wiki/Module:Bio::DB::SeqFeature::Store">Bio::DB::SeqFeature::Store</a>, or <a href="http://www.bioperl.org/wiki/Module:Bio::DB::GFF">Bio::DB::GFF</a>, then you can use JBrowse's <span class="code"><a href="#biodb-to-json.pl">biodb-to-json.pl</a></span>.  On the other hand, if you mainly deal with flat files like GFF or BED, it may be more convenient to use JBrowse's <span class="code"><a href="#flatfile-to-json.pl">flatfile-to-json.pl</a></span>.
      </p>
      <a name="biodb-to-json.pl" id="biodb-to-json.pl"><h4><span class="code">biodb-to-json.pl</span> -</h4></a>
      <p>
        You use <span class="code">bin/biodb-to-json.pl</span> with a configuration file <a href="../config.html">(documented here)</a>.
      </p>
      <pre>$ bin/biodb-to-json.pl --conf docs/tutorial/conf_files/volvox.json</pre>
      <a name="flatfile-to-json.pl" id="flatfile-to-json.pl"><h4><span class="code">bin/flatfile-to-json.pl</span> -</h4></a>
      <p>
        <span class="code">bin/flatfile-to-json.pl</span> doesn't use a configuration file; instead, per-track settings are given as command-line arguments.  Run <span class="code">bin/flatfile-to-json.pl</span> without any arguments to see a description of the available settings.  Most of them correspond to track settings in the config file, so see <a href="../config-desc.html#track_settings">the track settings section in the config file documentation</a> for more details.
      </p>
      <a name="Next-gen reads" id="Next-gen"></a>
      <h3>Next-gen reads</h3>
      <p>
        If you have data in the BAM format, you can import it into a JBrowse instance using the flatfile-to-json.pl program.  Here is one example of a command that does that, using the included data:
      </p>
      <pre>bin/flatfile-to-json.pl --bam docs/tutorial/data_files/volvox-sorted.bam --tracklabel bam_simulated --key "Simulated next-gen reads" --cssClass basic --clientConfig '{"featureCss": "background-color: #66F; height: 8px", "histCss": "background-color: #88F"}'</pre>
      <a name="Names" id="Names"></a>
      <h3>Names</h3>
      <p>
        To enable users to search using feature names or IDs, you have to do two things:
        <ol>
          <li>specify <span class="code">autocomplete</span> for the tracks whose features you want to be searchable (see <a href="../config-desc.html#track_settings">the track settings section in the config file documentation</a> for details about the <span class="code">autocomplete</span> setting).  When you do this, a file with a list of names and locations will be generated for each track that has the <span class="code">autocomplete</span> setting.</li>
          <li>run <span class="code">bin/generate-names.pl</span> to collect those per-track files together into the structure that JBrowse uses to search</li>
        </ol>
      </p>
      <pre>$ bin/generate-names.pl -v</pre>
      <p>
        You need to run <span class="code">bin/generate-names.pl</span> again every time you add new features to JBrowse.
      </p>

      <h4>And that's all, for basic usage.</h4>
      <p>visit <a href="../../index.html">the main page</a> to view the results.</p>

      <a name="Quantitative_tracks" id="Quantitative_tracks"></a>
      <h3>Quantitative tracks</h3>
      <p>
        If you want to browse quantitative ("wiggle") tracks in JBrowse, you need to do a bit more work.
      </p>
      <p>
        Prerequisites:
        <ul>
          <li>libpng</li>
        </ul>
      </p>
      <p>
        Compile the wiggle-processing program by running <span class="code">configure</span> and then <span class="code">make</span> in the JBrowse root directory
        <pre>$ ./configure
$ make</pre>
      </p>
      <div class="specific">
        <i>System-specific:</i>
        <ul>
          <li>
            Apple OS X:<br>
            <p>You need libpng and a C++ compiler; you can get these from <a href="http://www.macports.org/">MacPorts</a> or <a href="http://www.finkproject.org/">Fink</a>.</p>
            <p>Once you have those installed, you need to compile the JBrowse wiggle-processing program.  JBrowse includes a makefile to do this, but you may need to add the paths for <span class="code">libpng</span> and <span class="code">png.h</span> to your compiler's library and include paths. For example, if libpng is in <span class="code">/usr/X11</span>, you can run <span class="code">configure</span> like this:</p>
            <pre>$ ./configure CXXFLAGS=-I/usr/X11/include LDFLAGS=-L/usr/X11/lib</pre>
            <p>Depending on which OS X you have (and whether you're using MacPorts/fink/etc.), libpng might be somewhere other than <span class="code">/usr/X11</span>, like <span class="code">/opt/local</span> for example.  To find where libpng is, try <span class="code">locate libpng</span></p>
          </li>
          <br>
          <li>
            Red Hat/CentOS/Fedora:<br>
            install the <span class="code">libpng</span> and <span class="code">libpng-devel</span> packages.  (The rest should just work)
          </li>
          <br>
          <li>
            Ubuntu:<br>
            install the <span class="code">libpng12</span> and <span class="code">libpng12-dev</span> packages.  (The rest should just work)
        </ul>
      </div>
      <p>
        Once wig2png is compiled, you run it via a wrapper script:
      </p>
      <pre>$ bin/wig-to-json.pl --wig docs/tutorial/data_files/volvox_microarray.wig</pre>
    </div>
  </body>
</html>
